include/crm_internal.h

```diff
@@ -147,6 +147,22 @@ realloc_safe(void *ptr, size_t size)
     return new_ptr;
 }
 
+static inline GString *
+pcmk__str_append(GString *buffer, const gchar *format, ...)
+{
+    va_list args;
+
+    if (!buffer) {
+        buffer = g_string_new(NULL);
+    }
+
+    va_start(args, format);
+    g_string_append_vprintf(buffer, format, args);
+    va_end(args);
+
+    return buffer;
+}
```

Comment:
This would go in include/crm/common/internal.h (under "string functions"). There's no hard and fast rule on that, but we're moving toward crm_internal.h just being a wrapper that includes all the other internal headers.

For the question of the return value, some general guidelines: for a simple function that can't fail, returning a result (vs success/failure code) is fine. For functions that can fail, it's better to pass the result as an output argument and return a success/failure code (in pacemaker's case, a standard Pacemaker return code i.e. pcmk_rc_* or errno).

However a function like this, where the result is also one of the arguments, is a bit special. It's case by case, but we should prefer an input/output argument. The reason is that static analysis doesn't like if you set a variable to a return value and then never look at that variable again. That shouldn't happen for this particular case but it's better to avoid the possibility up front.

However however :) I'm not sure this particular function is a good idea. I'd rather callers call g_string_new() (or g_string_sized_new()) explicitly themselves, and then use glib functions like g_string_append() or g_string_printf(). The reasons are: calling g_string_new() will be a reminder that g_string_free() must also be called; g_string_sized_new() is much more efficient when the caller has an idea of a typical size (otherwise glib ends up doing many small reallocs); and glib offers a variety of functions for adding to a string and different ones might be more appropriate in different cases.

Reply:
> This would go in include/crm/common/internal.h (under "string functions").

Ack, good to know. Last night I saw a whole bunch of header files, couldn't immediately tell where this would go, and just picked one that gets included everywhere.

> we should prefer an input/output argument.

That all makes sense :) I thought better of the return value scheme after I signed off and took a breather.

> I'd rather callers call g_string_new() (or g_string_sized_new()) explicitly themselves,

Works for me. There's not that much added convenience from a `pcmk__str_append`-like function anyway.
I thought about `g_string_sized_new()`... maybe at the time I told myself I'd evaluate that later, or thought it wasn't worth complicating `pcmk__str_append()`. We could reuse the existing buffer sizes when creating a string that way. They seem to be appropriate or at least sufficient most of the time, and wouldn't waste that much memory.

Reply:
This function also produced a format string vulnerability warning on clang, so now I no longer have to think about what to do with that.
```
../../include/crm_internal.h:160:37: error: format string is not a string

      literal [-Werror,-Wformat-nonliteral]

    g_string_append_vprintf(buffer, format, args);

                                    ^~~~~~
```